test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.Extensibility.TokenReplay.cs

```diff
@@ -0,0 +1,277 @@
+using Microsoft.IdentityModel.Tokens;
+using Xunit;
+
+#nullable enable
+namespace Microsoft.IdentityModel.JsonWebTokens.Extensibility.Tests
+{
+    public partial class JsonWebTokenHandlerValidateTokenAsyncTests
+    {
+        [Theory, MemberData(nameof(TokenReplay_ExtensibilityTestCases), DisableDiscoveryEnumeration = true)]
+        public async Task ValidateTokenAsync_TokenReplayValidator_Extensibility(TokenReplayExtensibilityTheoryData theoryData)
+        {
+            var context = TestUtilities.WriteHeader($"{this}.{nameof(ValidateTokenAsync_TokenReplayValidator_Extensibility)}", theoryData);
+            context.IgnoreType = false;
+            for (int i = 0; i < theoryData.ExtraStackFrames; i++)
+                theoryData.TokenReplayValidationError!.AddStackFrame(new StackFrame(false));
+
+            try
+            {
+                ValidationResult<ValidatedToken> validationResult = await theoryData.JsonWebTokenHandler.ValidateTokenAsync(
+                    theoryData.JsonWebToken!,
+                    theoryData.ValidationParameters!,
+                    theoryData.CallContext,
+                    CancellationToken.None);
+
+                if (validationResult.IsValid)
+                {
+                    context.Diffs.Add("validationResult.IsValid is true, expected false");
+                }
+                else
+                {
+                    ValidationError validationError = validationResult.UnwrapError();
+                    IdentityComparer.AreValidationErrorsEqual(validationError, theoryData.TokenReplayValidationError, context);
+                    theoryData.ExpectedException.ProcessException(validationError.GetException(), context);
+                }
+            }
+            catch (Exception ex)
+            {
+                theoryData.ExpectedException.ProcessException(ex, context);
+            }
+
+            TestUtilities.AssertFailIfErrors(context);
+        }
+
+        public static TheoryData<TokenReplayExtensibilityTheoryData> TokenReplay_ExtensibilityTestCases
+        {
+            get
+            {
+                var theoryData = new TheoryData<TokenReplayExtensibilityTheoryData>();
+                CallContext callContext = new CallContext();
+                var utcNow = DateTime.UtcNow;
+                var expirationTime = utcNow + TimeSpan.FromHours(1);
+
+                #region return CustomTokenReplayValidationError
+                // Test cases where delegate is overridden and return a CustomTokenReplayValidationError
+                // CustomTokenReplayValidationError : TokenReplayValidationError, ExceptionType: SecurityTokenReplayDetectedException
+                theoryData.Add(new TokenReplayExtensibilityTheoryData(
+                    "CustomTokenReplayValidationDelegate",
+                    utcNow,
+                    CustomTokenReplayValidationDelegates.CustomTokenReplayValidationDelegate,
+                    extraStackFrames: 2)
```

Comment:
I added it simply because the 2 alone does not provide much context and this prevents having to go to the method definition to understand what the number means.